File: solutions/n_1162.py
def main():
    data = input().split()
    it = iter(data)
    n = int(next(it))
    m = int(next(it))
    S = int(next(it)) - 1
    V = float(next(it))
    amount = [0.0] * n
    amount[S] = V

    # читаем описание пунктов обмена
    exch = []
    for _ in range(m):
        A = int(next(it)) - 1
        B = int(next(it)) - 1
        RAB = float(next(it))
        CAB = float(next(it))
        RBA = float(next(it))
        CBA = float(next(it))
        exch.append((A, B, RAB, CAB, RBA, CBA))

    eps = 1e-12
    # relax–итерации (m раз достаточно, потому что простая последовательность не длиннее m)
    for _ in range(m):
        updated = False
        for A, B, RAB, CAB, RBA, CBA in exch:
            # обмен A -> B
            if amount[A] > CAB + eps:
                gained = (amount[A] - CAB) * RAB
                if gained > amount[B] + eps:
                    amount[B] = gained
                    updated = True
            # обмен B -> A
            if amount[B] > CBA + eps:
                gained = (amount[B] - CBA) * RBA
                if gained > amount[A] + eps:
                    amount[A] = gained
                    updated = True
        # если уже увеличили исходную валюту — достаточно
        if amount[S] > V + eps:
            print("YES")
            return
        if not updated:
            break

    # финальная проверка на «улучшающий цикл»
    for A, B, RAB, CAB, RBA, CBA in exch:
        if amount[A] > CAB + eps and (amount[A] - CAB) * RAB > amount[B] + eps:
            print("YES")
            return
        if amount[B] > CBA + eps and (amount[B] - CBA) * RBA > amount[A] + eps:
            print("YES")
            return

    print("NO")


if __name__ == "__main__":
    main()

File: solutions/n_1162.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

const eps = 1e-9

type Edge struct {
	from, to  int
	rate, com float64
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	var N, M, S int
	var V float64

	fmt.Fscan(reader, &N, &M, &S, &V)

	edges := make([]Edge, 0, 2*M)
	for i := 0; i < M; i++ {
		var A, B int
		var R_ab, C_ab, R_ba, C_ba float64
		fmt.Fscan(reader, &A, &B, &R_ab, &C_ab, &R_ba, &C_ba)
		edges = append(edges, Edge{from: A, to: B, rate: R_ab, com: C_ab})
		edges = append(edges, Edge{from: B, to: A, rate: R_ba, com: C_ba})
	}

	dist := make([]float64, N+1)
	dist[S] = V

	for i := 0; i < N; i++ {
		updated := false
		for _, e := range edges {
			if dist[e.from] > e.com+eps {
				amt := (dist[e.from] - e.com) * e.rate
				if amt > dist[e.to]+eps {
					dist[e.to] = amt
					updated = true
					if i == N-1 {
						fmt.Println("YES")
						return
					}
				}
			}
		}
		if !updated {
			break
		}
	}

	if dist[S] > V+eps {
		fmt.Println("YES")
	} else {
		fmt.Println("NO")
	}
}
